### PR TITLE
Multi eukulele

### PR DESCRIPTION
**Pull request by danilodileo**
I modified this PR in order to get a nicer way to call eukulele when we decide to have multiple runs.

usage:
nextflow run main.nf -profile test,docker --eukulele_db phylodb,mmetsp --eukulele_dbpath your/path/db

this update should work with the new eukulele release with --database gtdb option

N.B. this PR cover issue #10 too

### CODE REVIEW

**Comment by erikrikarddaniel** [comment]
I think it would be better if this behaviour was triggered by `eukulele_db` being given a list, e.g. `--eukulele_db phylodb,mmetsp`.

File: conf/modules.config [hunk 1]
```diff
@@ -292,10 +292,9 @@ process {
         memory = 16.GB
         ext.args = [
             params.orf_caller == 'transdecoder' ? "--protein_extension .pep" : '',
-            params.eukulele_db ? "--database ${params.eukulele_db}" : '',
             params.eukulele_method ? "-m ${params.eukulele_method}" : ''
         ].join(' ').trim()
-    }
+    ''}
```
Comment (erikrikarddaniel) [inline conf/modules.config]: Is there a newline missing?

File: modules/local/collect_featurecounts.nf [hunk 2]
```diff
@@ -53,7 +53,7 @@ process COLLECT_FEATURECOUNTS {
         ) %>%
         tidyr::unnest(d) %>%
         select(-f) %>%
-        write_tsv("counts.tsv.gz")
+        write_tsv("${meta}_counts.tsv.gz")
```
Comment (erikrikarddaniel) [inline modules/local/collect_featurecounts.nf]: Should be `meta.id`. You're probably passing in a `meta` that is just a name and not an object.

File: modules/local/eukulele/download.nf [hunk 3]
```diff
@@ -8,19 +8,23 @@ process EUKULELE_DB {
         'quay.io/biocontainers/eukulele:2.0.3--pyh723bec7_0' }"
 
     input:
+    val(db)
 
     output:
     path "versions.yml", emit: version
-    path("phylodb/")   , emit: phylo_db , optional: true
-    path("mmetsp/")    , emit: mmetsp_db, optional: true
+    path("${db}")      , emit: db , optional: true
```
Comment (erikrikarddaniel) [inline modules/local/eukulele/download.nf]: Should not be optional any more.

File: modules/local/eukulele/download.nf [hunk 4]
```diff
@@ -8,19 +8,23 @@ process EUKULELE_DB {
...
 
     script:
     def args = task.ext.args ?: ''
 
     """
     EUKulele \\
     download \\
-    $args
+    $args \\
+    --database $db
```
Comment (erikrikarddaniel) [inline modules/local/eukulele/download.nf]: Indent one level.

File: modules/local/eukulele/main.nf [hunk 5]
```diff
@@ -8,28 +8,34 @@ process EUKULELE {
         'quay.io/biocontainers/eukulele:2.0.3--pyh723bec7_0' }"
 
     input:
-    tuple val(meta), path(contigs_fasta)
-    path(db)
+    tuple val(meta)  , path(fasta)
+    tuple path(eukdb), val(namedb)
```
Comment (erikrikarddaniel) [inline modules/local/eukulele/main.nf]: Would it be better to have `tuple val(meta), path(fasta), val(dbname), path(eukdb)`? When you call the `EUKULELE` module with directory input (no name), pass in a `[]` for the `val(dbname)`.

File: subworkflows/local/eukulele.nf [hunk 6]
```diff
@@ -2,40 +2,39 @@
...
-            } else
-                EUKULELE(fastaprot, EUKULELE_DB.out.phylo_db)
-                FORMAT_TAX(EUKULELE.out.taxonomy_estimation.map { it[1] } )
-        } else {
-                ch_database = Channel.fromPath(params.eukulele_dbpath)
-                EUKULELE(fastaprot, ch_database)
-                FORMAT_TAX(EUKULELE.out.taxonomy_estimation.map { it[1] } )
-        }
+            EUKULELE_DB(eukulele_db)
+            EUKULELE(fastaprot,EUKULELE_DB.out.db, eukulele_db)
```
Comment (erikrikarddaniel) [inline subworkflows/local/eukulele.nf]: Alternative: Combine all three channels into one and modify `EUKULELE` so it takes only one channel, see above. Requires use of `combine()` to create a channel that is the cartesian product of faa (always just one) and the eukulele databases (one or more).

File: modules/local/eukulele/main.nf [hunk 7]
```diff
@@ -8,28 +8,34 @@ process EUKULELE {
...
+    gunzip   = fasta =~ /\.gz$/ ? "gunzip -c ${fasta} > ${input}" : ""
 
     """
+
+    $gunzip
+
     rc=0
     mkdir contigs
-    cp ${contigs_fasta} ./contigs/
+    cp $input ./contigs/
```
Comment (erikrikarddaniel) [inline modules/local/eukulele/main.nf]: `mv`

**Review by erikrikarddaniel — COMMENTED** [review]
A few comments.

File: conf/modules.config (hunk 1 above)
Comment (danilodileo) [reply conf/modules.config]: I removed the ''

File: modules/local/collect_featurecounts.nf (hunk 2 above)
Comment (danilodileo) [reply modules/local/collect_featurecounts.nf]: solved

File: modules/local/eukulele/download.nf (hunk 3 above)
Comment (danilodileo) [reply modules/local/eukulele/download.nf]: solved

File: modules/local/eukulele/download.nf (hunk 4 above)
Comment (danilodileo) [reply modules/local/eukulele/download.nf]: done

File: modules/local/eukulele/main.nf (hunk 5 above)
Comment (danilodileo) [reply modules/local/eukulele/main.nf]: done

File: workflows/metatdenovo.nf [hunk 8]
```diff
@@ -51,6 +52,18 @@ if ( params.hmmdir ) {
...
 
+ch_eukulele_db = Channel.empty()
+if ( !params.skip_eukulele ) {
+    if ( params.eukulele_db ) {
+        Channel
+            .of ( params.eukulele_db.split(',') )
+            .set { ch_eukulele_db }
+    } else {
+        ch_eukulele_db = []
+        }
```
Comment (danilodileo) [inline workflows/metatdenovo.nf:63]: indent

File: workflows/metatdenovo.nf [hunk 9]
```diff
@@ -423,12 +418,16 @@ workflow METATDENOVO {
...
 
     if( !params.skip_eukulele){
-        // DL: I think this should also be change so that the Eukulele module does the unzipping itself.
-        if ( params.orf_caller == ORF_CALLER_PROKKA ) {
-            UNPIGZ_EUKULELE(ch_aa)
-            SUB_EUKULELE(UNPIGZ_EUKULELE.out.unzipped.collect { [ [ id: 'all_samples' ], it ] } )
-        } else
-        SUB_EUKULELE(ch_aa)
+        if ( !params.eukulele_db ) {
+            SUB_EUKULELE( ch_aa.map {[ [ id:'all_samples'], it[1], [] ] } )
```
Comment (danilodileo) [inline workflows/metatdenovo.nf]: add assembly.orf_caller

**Comment by danilodileo** [comment]
The module has some features that needs to be fixed.
1. eukulele can't run with a custom database for taxonomic annotation
2. If you ask a double run, the sub workflow will run eukulele as many combination it can create from the .map output. for example with two database selected (--eukulele_db phylodb,mmetsp) it will try to run eukulele 4 times. 
3. it is not possible at to avoid eukulele_download module. that is an issue for the previous two points

**Review by erikrikarddaniel — APPROVED** [review]
LGTM.

**Pull request merged**